extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,608 @@
+      if (task instanceof KafkaIndexTask && dataSource.equals(task.getDataSource())) {
+        taskCount++;
+        KafkaIndexTask kafkaTask = (KafkaIndexTask) task;
+
+        // determine which task group this task belongs to and do a consistency check on partitions
+        Integer taskGroupId = null;
+        for (Integer partition : kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet()) {
+          if (taskGroupId == null) {
+            taskGroupId = getTaskGroupIdForPartition(partition);
+          } else if (!taskGroupId.equals(getTaskGroupIdForPartition(partition))) {
+            log.warn("Killing task [%s] which does not match the expected partition allocation", task.getId());
+            killTask(task.getId());
+            taskGroupId = null; // prevents the next block of code from adding the item to taskGroups
+            break;
+          }
+        }
+
+        if (taskGroupId != null) {
+          if (taskGroups.get(taskGroupId) == null) {
+            log.info("Creating new task group [%d]", taskGroupId);
+            taskGroups.put(taskGroupId, new TaskGroup());
+          }
+
+          if (!taskGroups.get(taskGroupId).taskStatuses.containsKey(task.getId())) {
+            taskGroups.get(taskGroupId).taskStatuses.put(task.getId(), Optional.<TaskStatus>absent());
+          }
+        }
+      }
+    }
+
+    // update TaskStatus for tracked tasks
+    for (Integer key : taskGroups.keySet()) {
+      for (String taskId : taskGroups.get(key).taskStatuses.keySet()) {
+        taskGroups.get(key).taskStatuses.put(taskId, taskStorage.getStatus(taskId));
+      }
+    }
+
+    log.debug("Found [%d] Kafka indexing tasks for dataSource [%s]", taskCount, dataSource);
+  }
+
+  private void checkTaskState()
+  {
+    for (Integer groupId : taskGroups.keySet()) {
+      TaskGroup taskGroup = taskGroups.get(groupId);
+
+      // Iterate the list of known tasks in this group and:
+      //   1) Kill any tasks which are not "current" (have the same partitions and offset ranges as the latest
+      //      known set.
+      //   2) Determine if any task completed successfully.
+      //   3) Remove any tasks that have failed from the list.
+      //
+      // When we're done this loop, we will either have:
+      //   1) Found a successful current task, which will trigger a shutdown of the remaining running replicas and will
+      //      queue the next set of tasks.
+      //   2) Not found a successful task, and have a taskStatuses list that only contains current tasks that are in
+      //      running state. We will then compare the number of running tasks to the expected number of replicas and
+      //      create additional tasks as necessary.
+
+      log.debug("Task group [%d] pre-pruning: %s", groupId, taskGroup.taskStatuses);
+      String successfulTaskId = null;
+      Iterator<Map.Entry<String, Optional<TaskStatus>>> it = taskGroup.taskStatuses.entrySet().iterator();
+      while (it.hasNext()) {
```

Comment:
@pjain1 removing the TaskGroup when all tasks are killed would be okay; the only (minor) downside is that with the current logic the new retry tasks wouldn't be enqueued until the next cycle. However, I don't believe this is necessary since Kafka does not support decreasing the number of partitions, and I can't think of any other reason that the number of task groups would decrease.